### DISCUSSION TITLE
Add a 'physical' provider

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#99] [Ideas]
It can be useful to validate a given topology using virtual nodes, and then apply the same configs to physical nodes.

This could be accomplished by having a 'physical' provider, which would assume there is a physical node reachable at the given management address. This provider would include the logic required to transfer generated configs to physical devices, and possibly the ability to build hybrid setups (mix of virtual and physical nodes)

**Comment by ipspace** [comment]
I totally agree it would be useful, but there are tons of hurdles. We could start with something minimal like _using this provider means we're not doing any stuff_. Even there:

* You have to be able to specify management IP in the nodes and make it mandatory with _physical_ provider (minor code changes, doable)
* You have to be able to specify real interface names in node-on-link (interface) data. Yet again, minor code changes.

The "only" problem with "minor code changes" is the exploding decision tree. Just look at how complex `augment.links.get_node_link_address` got.

I did work on physical lab provisioning tools in the past, and the true problem is pushing initial device configurations and cleaning up device configs, in particular once the device gets bricked. That problem would have to be ignored (we're not building _that_ product).

Hybrid setup is a whole different can of worms. I'm thinking about mixing VMs with containers in _libvirt_; once I get that done we'd have at least the minimal multi-provider infrastructure in place. The gluing part needed to get the hybrid networking working is way beyond the scope of this tool though.

To recap:

* Interesting idea
* Firmly on the backburner from my personal perspective ;)
* If anyone feels like throwing together a proof-of-concept with some usable code please go for it.

**Comment by ipspace** [comment]
Implemented in 'external' provider
